amplipi/mdns.py

```diff
@@ -0,0 +1,179 @@
+        return ip_addr, mac_addr, iface
+    except:
+      pass
+    sleep(2)  # wait a bit in case this was started before DHCP was started
+    retry_count -= 1
+
+  # fallback to any normal interface that is available
+  # this also covers the case where a link-local connection is established
+  def is_normal(iface: str):
+    """ Check if an interface is wireless or wired, excluding virtual and local interfaces"""
+    return iface.startswith('w') or iface.startswith('e')
+  ifaces = filter(is_normal, ni.interfaces())
+  for iface in ifaces:
+    ip_addr, mac_addr = get_ip_info(iface)
+    if ip_addr and mac_addr:
+      return ip_addr, mac_addr, iface
+  log.warning(f'Unable to register service on one of {ifaces}, \
+          they all are either not available or have no IP address.')
+  return None, None, None
+
+
+def _advertise_service(port: int, ok: Callable) -> None:
+  """ Underlying advertisement, can throw Exceptions when network is not connected """
+
+  hostname = f'{gethostname()}.local'
+  url = f'http://{hostname}'
+  log.debug("AmpliPi zeroconf - attempting to find best interface")
+  ip_addr, mac_addr, good_iface = _find_best_iface(ok)
+
+  if not ip_addr:
+    log.warning('is this running on AmpliPi?')
+    # Fallback to any hosted ip on this device.
+    # This will be resolved to an ip address by the advertisement
+    ip_addr = '0.0.0.0'
+  else:
+    log.debug('Found IP address %s on interface %s', ip_addr, good_iface)
+  if port != 80:
+    url += f':{port}'
+
+  info = ServiceInfo(
+    # use a custom type to easily support multiple amplipi device enumeration
+    '_amplipi._tcp.local.',
+    # the MAC Address is appended to distinguish multiple AmpliPi units
+    f'amplipi-{str(mac_addr).lower()}._amplipi._tcp.local.',
+    addresses=[inet_aton(ip_addr)],
+    port=port,
+    properties={
+      # standard info
+      'path': '/api/',
+      # extra info - for interfacing
+      'name': 'AmpliPi',
+      'vendor': 'MicroNova',
+      'version': utils.detect_version(),
+      # extra info - for user
+      'web_app': url,
+      'documentation': f'{url}/doc'
+    },
+    server=f'{hostname}.',  # Trailing '.' is required by the SRV_record specification
+  )
+
+  if not ok():
+    log.info("Advertisement cancelled")
+    return
+
+  log.info(f'Registering service: {info}')
+  # right now the AmpliPi webserver is ipv4 only
+  zeroconf = Zeroconf(ip_version=IPVersion.V4Only, interfaces=[ip_addr])
+  zeroconf.register_service(info)
+  log.info('Finished registering service')
+  try:
+    # poll for changes to the IP address
+    # this attempts to handle events like router/switch resets
+    while ok():
+      delay_ct = 100
+      while ok() and delay_ct > 0:
+        sleep(0.1)
+        delay_ct -= 1
+      if ok():
+        new_ip_addr, _, new_iface = _find_best_iface(ok)
+        if new_ip_addr != ip_addr:
+          log.info(f'IP address changed from {ip_addr} ({good_iface}) to {new_ip_addr} ({new_iface})')
```

Comment:
pylint is warning about f-strings in logging. Seems like the TLDR is it's more performant to pass log.info('Message %s', variable) since then if the logging level is changed the f-string doesn't have to be evaluated. It also means if an error occurs during evaluation of the f-string it gets logged. I suspect these aren't good enough reasons for us to have a second method of string formatting to be remembered, so I added `disable=logging-fstring-interpolation` to our .pylintrc. Feel free to remove this override and fix the logging strings if you feel otherwise. See here for some of the discussion on this: https://github.com/pylint-dev/pylint/issues/1788

Reply:
> so I added `disable[...]`

:clap: